accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,493 @@
+
+The member declarations can be any static or instance member declaration except for instance variables and constructors.
+
+### Scope
+
+Dart's static extension methods are *scoped*. They only apply to code that the extension itself is accessible to. Being accessible means that using the extension's name must denote the extension. The extension is not in scope if another declaration with the same name shadows the extension, if the extension's library is not imported, if the library is imported and the extension is hidden, or the library is only imported with a prefix. In other words, if the extension had been a class, it is only in scope if using the name would denote the class&mdash; which will allow you to call static methods on the class, which is exactly what we are going to do.
+
+### Extension Member Resolution
+
+The declaration introduces an extension. The extension's `on` type defines which types are being extended.
+
+For any member access, `x.foo`, `x.bar()`, `x.baz = 42`, `x(42)` or `x + y`, the language first checks whether the static type of `x` has a member with the same base name as the operation. That is, if it has a corresponding instance member, respectively, a `foo` method or getter or a `foo=` setter. a `bar` member or `bar=` setter, a `baz` member or `baz=` setter, a `call` method, or a `+` method. If so, then the operation is unaffected by extensions. *This check does not care whether the invocation is otherwise correct, based on number or type of the arguments, it only checks if there is a member at all.*
+
+(The types `dynamic` and `Never` are considered as having all members, the type `void` is always a compile-time error when used in a receiver position, so none of these can ever be affected by static extension methods).
+
+If there is no such member, the operation is currently a compile-time error. In that case, all extensions in scope are checked for whether they apply. An extension applies to a member access if the static type of the receiver is a subtype of the `on` type of the extension *and* the extension has an instance member with the same base name as the operation. 
+
+For generic extensions, standard type inference is used to infer the type arguments. As an example, take the following extension:
+
+```dart
+extension MyList<T extends Comparable<T>> on List<T> {
+  void quickSort() { ... }
+}
+```
+
+and a member access like:
+
+```dart
+Uint8List bytes = ...;
+bytes.quickSort();
+```
+
+Here we perform type inference equivalent to what we would do for:
+
+```dart
+class MyList<T extends Comparable<T>> {
+  MyList(List<T> argument);
+  void quickSort() { ... }
+}
+...
+Uint8List bytes = ...;
+... MyList(bytes).quickSort() ... // Infer type argument of MyList here.
```

Comment:
You should make clear that the context type for the invocation is not used as part of this inference process.